File: src/BubbleSort.java
import java.util.Arrays;

/**
 * Created by konstantin.silin on 23.10.2015.
 */
public class BubbleSort {

    public static void main(String[] args) {
        int[] arrToSort = Utils.getRandomIntArray(10);
        System.out.println(Arrays.toString(arrToSort));
        BubbleSort sorter = new BubbleSort();
        sorter.sort(arrToSort);
        System.out.println(Arrays.toString(arrToSort));

        System.out.println("Check sorted array: " + Utils.checkSort(arrToSort));
    }

    private void sort(int[] arrToSort) {
        if (arrToSort.length > 1) {
            int buff = 0;
            for (int j = 0; j < arrToSort.length; j ++) {
                boolean isSorted = true;
                for (int i = 0; i < arrToSort.length - 1; i++) {
                    if (arrToSort[i] > arrToSort[i + 1]) {
                        isSorted = false;
                        buff = arrToSort[i];
                        arrToSort[i] = arrToSort[i + 1];
                        arrToSort[i + 1] = buff;
                    }
                }

                if (isSorted) {
                    System.out.println("Preliminary sorted");
                    break;
                }
            }
        }
    }
}

File: sorting/src/main/kotlin/ru/ksilin/study/sorting/BubbleSort.kt
package ru.ksilin.study.sorting

import ru.ksilin.study.sorting.Utils.swap

fun bubbleSort(arr: IntArray) {
    if (arr.isEmpty()) {
        return
    }

    for (i in arr.lastIndex downTo 0) {
        for (j in 0 until i) {
            if (arr[j] > arr[j+1]) {
                swap(arr, j, j+1)
            }
        }
    }
}

fun bubbleSortTwoWay(arr: IntArray) {
    var rightIdx = arr.lastIndex
    var leftIdx = 0
    while (rightIdx > leftIdx) {
        for (i in leftIdx until rightIdx) {
            if (arr[i] > arr[i+1])
                swap(arr, i, i+1)
        }
        --rightIdx
        for (i in rightIdx downTo leftIdx+1) {
            if (arr[i] < arr[i-1])
                swap(arr, i, i-1)
        }
        ++leftIdx
    }
}

fun oddEvenSort(arr: IntArray) {
    var isSorted = false
    while (!isSorted) {
        isSorted = true
        for (i in 0 until arr.lastIndex step 2) {
            if (arr[i] > arr[i+1]) {
                swap(arr, i, i+1)
                isSorted = false
            }
        }

        for (i in 1 until arr.lastIndex step 2) {
            if (arr[i] > arr[i+1]) {
                swap(arr, i, i+1)
                isSorted = false
            }
        }
    }
}

fun main() {
//    checkBubbleSort()
//    checkTwoWayBubbleSort()
    checkOddEvenSort()
}

fun runSortFunction(sortFunction: (arr: IntArray) -> Unit) {
    val arrToSort = Utils.getRandomIntArray(10)
    val clone = arrToSort.clone()
    println(arrToSort.contentToString())
    arrToSort.sort()
    sortFunction(clone)
    println(arrToSort.contentToString())

    println("Check sorted array: " + Utils.checkSort(arrToSort))
    println("Are equal result in different methods = ${arrToSort.contentEquals(clone)}")
}

fun checkTwoWayBubbleSort() {
    runSortFunction { bubbleSortTwoWay(it) }
}

fun checkBubbleSort() {
    runSortFunction { bubbleSort(it) }
}

fun checkOddEvenSort() {
    runSortFunction { oddEvenSort(it) }
}